scrape/metadata_watcher.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scrape
+
+import (
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+)
+
+// MetadataAppender is an interface used by the Metadata Watcher to send metadata read from the scrape manager on to somewhere else.
+type MetadataAppender interface {
+	AppendMetadata([]MetricMetadata) bool
+}
+
+// MetadataWatcher watches the Scrape Manager for a given WriteMetadataTo.
+type MetadataWatcher struct {
+	name   string
+	logger log.Logger
+
+	managerGetter ManagerGetter
+	manager       *Manager
+	writer        MetadataAppender
+
+	interval model.Duration
+
+	quit chan struct{}
+	done chan struct{}
+}
+
+// NewMetadataWatcher builds a new MetadataWatcher.
+func NewMetadataWatcher(l log.Logger, mg ManagerGetter, name string, w MetadataAppender, interval model.Duration) *MetadataWatcher {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+	return &MetadataWatcher{
+		name:   name,
+		logger: l,
+
+		managerGetter: mg,
+		writer:        w,
+
+		interval: interval,
+
+		quit: make(chan struct{}),
+		done: make(chan struct{}),
+	}
+}
+
+// Start the MetadataWatcher.
+func (mw *MetadataWatcher) Start() {
+	level.Info(mw.logger).Log("msg", "starting scrapped metadata watcher")
+	go mw.loop()
+}
+
+// Stop the MetadataWatcher.
+func (mw *MetadataWatcher) Stop() {
+	close(mw.quit)
+	<-mw.done
+
+	level.Info(mw.logger).Log("msg", "scrapped metadata watcher stopped")
+}
+
+func (mw *MetadataWatcher) loop() {
+	ticker := time.NewTicker(time.Duration(mw.interval))
+	defer ticker.Stop()
+	defer close(mw.done)
+
+	for {
+		select {
+		case <-mw.quit:
+			return
+		case <-ticker.C:
+			mw.collect()
+		}
+	}
+}
+
+func (mw *MetadataWatcher) collect() {
+	if !mw.readyForCollection() {
+		return
+	}
+
+	metadata := []MetricMetadata{}
+	for _, targets := range mw.manager.TargetsActive() {
+		for _, tt := range targets {
+			metadata = append(metadata, tt.MetadataList()...)
+		}
+	}
+
+	// Blocks until the metadata is sent to all remote write endpoints or closed.
```

Comment:
is this right?

Reply:
I believe so, it is no longer parallel.

Reply:
But don't we have a watcher per remote write endpoint/queue manager?

Reply:
🤦‍♂ right! Reworded, looks better now?